# **Libraries**

In [1]:
import pandas as pd
import numpy as np

from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split

import lightgbm as lgb
from sklearn.metrics import roc_auc_score

from feature_engine.encoding import WoEEncoder
from feature_engine.imputation import CategoricalImputer
from feature_engine.imputation import ArbitraryNumberImputer

import functions
import importlib
importlib.reload(functions)

import warnings

# **Display**

In [2]:
%matplotlib inline

pd.options.display.max_rows = 300000
pd.options.display.max_columns = 999
pd.options.display.max_colwidth = 500

warnings.filterwarnings("ignore")
warnings.simplefilter(action="ignore", category=FutureWarning)

pd.set_option('display.max_rows', 200)

# **Load Data**

In [14]:
train = pd.read_csv(
    r"C:\Users\Dell\Documents\AI\Risk\Data\Data\train 22.csv",
    index_col=False
)

## **Reduce Memory Usage**

In [16]:
train = functions.reduce_memory_usage(train)

Memory usage of dataframe is 293.26 MB
Memory usage after optimization is: 99.71 MB
Decreased by 66.0%


## **Variables**

In [17]:
random_state = 101
target = 'TARGET'

## **Imputation**

In [18]:
ani = ArbitraryNumberImputer(arbitrary_number=-99999)
ani.fit(train)
train = ani.transform(train)

In [19]:
ci = CategoricalImputer(imputation_method='missing', fill_value='UNKNOWN')
ci.fit(train)
train = ci.transform(train)

## **WoE Encoder**

In [20]:
woe = WoEEncoder(fill_value=0.0001)
woe.fit(train, train[target])
train = woe.transform(train)

## **Train Test Split**

In [21]:
X = train.drop('TARGET', axis=1)
y = train['TARGET']

X, y = shuffle(X, y, random_state=random_state)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=random_state)

### **LGBM**

In [22]:
model = lgb.LGBMClassifier(boosting_type='gbdt', 
                           num_leaves=31, 
                           max_depth=-1, 
                           learning_rate=0.1, 
                           n_estimators=100,
                           verbose=-1)

model.fit(X_train, y_train)

y_prob = model.predict_proba(X_test)[:, 1]  

auc_score = roc_auc_score(y_test, y_prob)
print(f"AUC Score: {auc_score:.2f}")

AUC Score: 0.76


## **Feature Importance**

In [23]:
feature_importance = model.feature_importances_
feature_names = X.columns

importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': feature_importance})
importance_df = importance_df.sort_values(by='Importance', ascending=False).reset_index(drop=True)

importance_df

Feature  Importance
0           ANNUITY_TO_CREDIT_RATIO         206
1                   EXT_SOURCE_MEAN         160
2    ANNUAL_PAYMENT_TO_CREDIT_RATIO         140
3                               AGE         139
4                      EXT_SOURCE_3         128
5                  YEARS_ID_PUBLISH         114
6                       AMT_ANNUITY          97
7                   AMT_GOODS_PRICE          87
8                      EXT_SOURCE_2          82
9                      EXT_SOURCE_1          78
10          ANNUITY_TO_INCOME_RATIO          66
11                ORGANIZATION_TYPE          65
12                       AMT_CREDIT          62
13             YEARS_EMPLOYED_RATIO          62
14               YEARS_REGISTRATION          60
15          YEARS_LAST_PHONE_CHANGE          59
16       YEARS_EMPLOYED_AGE_PRODUCT          56
17                       SK_ID_CURR          55
18              INCOME_TO_AGE_RATIO          54
19              CREDIT_TO_AGE_RATIO          53
20               EXT_SOURCE_PRODUCT          50
21             DEBT_TO_INCOME_RATIO          47
22       REGION_POPULATION_RELATIVE          46
23                      OWN_CAR_AGE          43
24                      CODE_GENDER          41
25             CREDIT_PER_DEPENDENT          40
26         INCOME_PER_FAMILY_MEMBER          39
27                   YEARS_EMPLOYED          37
28                  OCCUPATION_TYPE          36
29              NAME_EDUCATION_TYPE          35
30          HOUR_APPR_PROCESS_START          33
31       AMT_REQ_CREDIT_BUREAU_YEAR          28
32                     FLAG_OWN_CAR          28
33               NAME_FAMILY_STATUS          28
34                 AMT_INCOME_TOTAL          27
35         DEF_30_CNT_SOCIAL_CIRCLE          27
36             INCOME_PER_DEPENDENT          24
37               NAME_CONTRACT_TYPE          24
38        AMT_REQ_CREDIT_BUREAU_QRT          23
39      REGION_RATING_CLIENT_W_CITY          21
40                  FLAG_DOCUMENT_3          18
41                  FLAG_WORK_PHONE          18
42       WEEKDAY_APPR_PROCESS_START          17
43         OBS_30_CNT_SOCIAL_CIRCLE          15
44                     LANDAREA_AVG          15
45                  APARTMENTS_MODE          14
46           REG_CITY_NOT_LIVE_CITY          13
47                NAME_HOUSING_TYPE          13
48         DEF_60_CNT_SOCIAL_CIRCLE          13
49                BASEMENTAREA_MODE          13
50                   TOTALAREA_MODE          12
51                 BASEMENTAREA_AVG          11
52                    LANDAREA_MODE          11
53               NONLIVINGAREA_MODE          11
54            LIVINGAPARTMENTS_MODE          11
55                  CNT_FAM_MEMBERS          11
56                   COMMONAREA_AVG          11
57                   APARTMENTS_AVG          11
58                  YEARS_BUILD_AVG          10
59                NONLIVINGAREA_AVG          10
60       AMT_REQ_CREDIT_BUREAU_HOUR          10
61     YEARS_BEGINEXPLUATATION_MODE          10
62      YEARS_BEGINEXPLUATATION_AVG           9
63     YEARS_BEGINEXPLUATATION_MEDI           9
64                   LIVINGAREA_AVG           8
65                  NAME_TYPE_SUITE           8
66                   ENTRANCES_MODE           8
67                 NAME_INCOME_TYPE           8
68                  COMMONAREA_MODE           8
69        AMT_REQ_CREDIT_BUREAU_MON           7
70                    ENTRANCES_AVG           7
71                  FLAG_OWN_REALTY           7
72                  LIVINGAREA_MODE           7
73            LIVINGAPARTMENTS_MEDI           7
74               NONLIVINGAREA_MEDI           7
75         OBS_60_CNT_SOCIAL_CIRCLE           6
76                       FLAG_PHONE           6
77                  LIVINGAREA_MEDI           6
78               WALLSMATERIAL_MODE           6
79             LIVINGAPARTMENTS_AVG           6
80                    FLOORSMAX_AVG           6
81        DEPENDENTS_TO_FAMILY_SIZE           5
82                  APARTMENTS_MEDI         

## **Keep Columns**

In [24]:
columns = ['SK_ID_CURR',
                   'ANNUITY_TO_CREDIT_RATIO',
                   'EXT_SOURCE_3',
                   'EXT_SOURCE_1',
                   'EXT_SOURCE_2',
                   'EXT_SOURCE_MEAN',
                   'ANNUAL_PAYMENT_TO_CREDIT_RATIO', 
                   'AGE', 
                   'YEARS_ID_PUBLISH',
                   'AMT_ANNUITY',
                   'AMT_GOODS_PRICE',
                   'ANNUITY_TO_INCOME_RATIO',
                   'ORGANIZATION_TYPE',
                   'AMT_CREDIT',
                   'YEARS_EMPLOYED_RATIO',
                   'YEARS_REGISTRATION',
                   'YEARS_LAST_PHONE_CHANGE',
                   'YEARS_EMPLOYED_AGE_PRODUCT',
                   'INCOME_TO_AGE_RATIO',
                   'CREDIT_TO_AGE_RATIO',
                   'EXT_SOURCE_PRODUCT',
                   'DEBT_TO_INCOME_RATIO',
                   'REGION_POPULATION_RELATIVE',
                   'OWN_CAR_AGE',
                   'CODE_GENDER',
                   'CREDIT_PER_DEPENDENT',
                   'TARGET'] 
train = train[columns]

In [25]:
train.shape

(307507, 27)

### **LGBM**

In [26]:
X = train.drop('TARGET', axis=1)
y = train['TARGET']

X, y = shuffle(X, y, random_state=random_state)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=random_state)

In [27]:
model = lgb.LGBMClassifier(boosting_type='gbdt', 
                           num_leaves=31, 
                           max_depth=-1, 
                           learning_rate=0.1, 
                           n_estimators=100,
                           verbose=-1)

model.fit(X_train, y_train)

y_prob = model.predict_proba(X_test)[:, 1]  

auc_score = roc_auc_score(y_test, y_prob)
print(f"AUC Score: {auc_score:.2f}")

AUC Score: 0.76


## **Feature Importance**

In [17]:
feature_importance = model.feature_importances_
feature_names = X.columns

importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': feature_importance})
importance_df = importance_df.sort_values(by='Importance', ascending=False).reset_index(drop=True)

importance_df

Feature  Importance
0          ANNUITY_TO_CREDIT_RATIO         316
1                  EXT_SOURCE_MEAN         194
2                              AGE         184
3   ANNUAL_PAYMENT_TO_CREDIT_RATIO         154
4                     EXT_SOURCE_3         152
5                 YEARS_ID_PUBLISH         138
6                     EXT_SOURCE_2         134
7                     EXT_SOURCE_1         123
8                  AMT_GOODS_PRICE         119
9          ANNUITY_TO_INCOME_RATIO         116
10             INCOME_TO_AGE_RATIO         110
11               ORGANIZATION_TYPE         109
12                     AMT_ANNUITY         104
13              YEARS_REGISTRATION         102
14         YEARS_LAST_PHONE_CHANGE         100
15                      SK_ID_CURR          91
16      REGION_POPULATION_RELATIVE          88
17                     OWN_CAR_AGE          87
18      YEARS_EMPLOYED_AGE_PRODUCT          84
19            CREDIT_PER_DEPENDENT          84
20            YEARS_EMPLOYED_RATIO          82
21             CREDIT_TO_AGE_RATIO          71
22                      AMT_CREDIT          70
23              EXT_SOURCE_PRODUCT          66
24            DEBT_TO_INCOME_RATIO          66
25                     CODE_GENDER          56

## **Drop Columns**

In [28]:
columns = ['CODE_GENDER',
           'OWN_CAR_AGE',
           'AMT_CREDIT',
           'YEARS_EMPLOYED_RATIO',
           'CREDIT_TO_AGE_RATIO',
           'DEBT_TO_INCOME_RATIO',
           'CREDIT_PER_DEPENDENT',
           'EXT_SOURCE_PRODUCT'
                ] 
train = train.drop(columns=columns)

In [29]:
train.shape

(307507, 19)

### **LGBM**

In [30]:
X = train.drop('TARGET', axis=1)
y = train['TARGET']

X, y = shuffle(X, y, random_state=random_state)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=random_state)

In [31]:
model = lgb.LGBMClassifier(boosting_type='gbdt', 
                           num_leaves=31, 
                           max_depth=-1, 
                           learning_rate=0.1, 
                           n_estimators=100,
                           verbose=-1)

model.fit(X_train, y_train)

y_prob = model.predict_proba(X_test)[:, 1]  

auc_score = roc_auc_score(y_test, y_prob)
print(f"AUC Score: {auc_score:.2f}")

AUC Score: 0.76


## **Feature Importance**

In [22]:
feature_importance = model.feature_importances_
feature_names = X.columns

importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': feature_importance})
importance_df = importance_df.sort_values(by='Importance', ascending=False).reset_index(drop=True)

importance_df

Feature  Importance
0          ANNUITY_TO_CREDIT_RATIO         374
1                  EXT_SOURCE_MEAN         219
2                              AGE         206
3   ANNUAL_PAYMENT_TO_CREDIT_RATIO         199
4                      AMT_ANNUITY         199
5                     EXT_SOURCE_3         182
6                  AMT_GOODS_PRICE         158
7       YEARS_EMPLOYED_AGE_PRODUCT         157
8                 YEARS_ID_PUBLISH         155
9                     EXT_SOURCE_2         154
10                    EXT_SOURCE_1         141
11         ANNUITY_TO_INCOME_RATIO         137
12             INCOME_TO_AGE_RATIO         133
13               ORGANIZATION_TYPE         123
14                      SK_ID_CURR         122
15              YEARS_REGISTRATION         118
16         YEARS_LAST_PHONE_CHANGE         117
17      REGION_POPULATION_RELATIVE         106

In [32]:
train.shape

(307507, 19)

# **Insights**

> * **Iterative Machine Learning** — There are approximately 300,000 rows and 122 columns. The final column is "TARGET" indicates if the client defaulted on a loan. There are enormous missing values in many of the columns. 
> * **Reduction of Number of Features** — Of the eight supplementary tables, I chose to use five for this project.
> * **Transparency** — Of the eight supplementary tables, I chose to use five for this project.

# **Save Dataframe as CSV File**

In [33]:
train.to_csv(r"C:\Users\Dell\Documents\AI\Risk\Data\Data\train 23.csv", index=False)